feat: add vm-migration-network setting (#811)
harvester/docs: docs/advanced/vm-migration-network.md

---
sidebar_position: 12
sidebar_label: VM Migration Network
title: "VM Migration Network"
---

A VM migration network is useful for isolating migration traffic from cluster traffic on `mgmt` and other cluster-wide workloads. Using a VM migration network results in better network bandwidth and performance.

:::note

Avoid configuring KubeVirt settings directly, as this can result in unexpected or unwanted system behavior.

:::

## Prerequisites

Before you begin configuring the VM migration network, ensure that the following requirements are met:

- The network switches are correctly configured, and a dedicated VLAN ID is assigned to the VM migration network.

- The [cluster network](../networking/clusternetwork.md) and [VLAN network](../networking/harvester-network.md) are configured correctly. Ensure that both networks cover all nodes and are accessible.

- No virtual machines are being migrated.

> - The `ippools.whereabouts.cni.cncf.io` CRD exists. You can check this using the command `kubectl get crd ippools.whereabouts.cni.cncf.io`. In certain [upgrade scenarios](https://github.com/harvester/harvester/issues/3168), the Whereabouts CNI is not installed correctly.

Reviewer: Not sure if we still need to mention this requirement, since the issue pertains to Harvester 1.x, which is EOL. For now, those CRDs should be installed by default in supported versions. May just remove this check. 

P.S. We may also need to update the storage network documentation. @jillian-maroket

> - The IP range of the VM migration network is in the IPv4 CIDR format and must neither conflict nor overlap with Kubernetes cluster networks. You must exclude IP addresses that KubeVirt pods and the VM migration network must not use. The following addresses are reserved: `10.42.0.0/16`, `10.43.0.0/16`, `10.52.0.0/16` and `10.53.0.0/16`.

Reviewer: You already explained the IP range for the VM migration network, but why is it necessary to mention "the VM migration network must not use"? It’s a little confusing here.

Reviewer 2: This is legacy text that was copied from the storage network doc and reorganized a bit. I recently updated the storage network doc. LMK if you want to reuse the wording that what Ivan and I agreed on:

<img width="879" height="690" alt="Screenshot 2025-08-07 at 5 09 41 PM" src="https://github.com/user-attachments/assets/5d8cf2f1-43a5-4336-8ea2-109d4424491f" />
https://docs.harvesterhci.io/v1.6/advanced/storagenetwork#prerequisites

:::tip

Using the Harvester UI to configure the `vm-migration-network` setting is strongly recommended.

:::

#### Enable the VM Migration Network

1. Go to **Advanced > Settings > vm-migration-network**.

1. Select **Enabled**.

1. Configure the **VLAN ID**, **Cluster Network**, **IP Range**, and **Exclude** fields to construct a Multus `NetworkAttachmentDefinition` CRD.

1. Click **Save**.

![vm-migration-network-enabled.png](/img/v1.6/vm-migration-network/vm-migration-network-enabled.png)

#### Disable the VM Migration Network

1. Go to **Advanced > Settings > vm-migration-network**.

1. Select **Disabled**.

1. Click **Save**.

Once the VM migration network is disabled, KubeVirt starts using `mgmt` for VM migration-related operations.

![vm-migration-network-disabled.png](/img/v1.6/vm-migration-network/vm-migration-network-disabled.png)

</TabItem>

<TabItem value="cli" label="CLI">

You can use the following command to configure the [`vm-migration-network`](./settings.md#vm-migration-network) setting.

```bash
kubectl edit settings.harvesterhci.io vm-migration-network
```

The VM migration network is automatically enabled in the following situations:

> - The value field contains a valid JSON string.

Reviewer 2: Feel free to use this if the existing wording is incorrect.

  ```yaml
  apiVersion: harvesterhci.io/v1beta1
  kind: Setting
  metadata:
    name: vm-migration-network
  value: '{"vlan":100,"clusterNetwork":"vm-migration","range":"192.168.1.0/24", "exclude":["192.168.1.100/32"]}'
  ```

> - The value field is empty.

Reviewer: Is it valid to have an empty value when the setting is enabled?

  ```yaml
  apiVersion: harvesterhci.io/v1beta1
  kind: Setting
  metadata:
    name: vm-migration-network
  ```

:::caution

Harvester considers extra insignificant characters in a JSON string as a different configuration.

:::

</TabItem>
</Tabs>

The following occur once the `vm-migration-network` setting is applied:

- Harvester creates a new `NetworkAttachmentDefinition` and updates the KubeVirt configuration.
- KubeVirt restarts all `virt-handler` pods to apply the new network configuration.

### Post-Configuration Steps

1. Verify that the setting's status is `True` and the type is `configured` using the following command:

    ```bash
    kubectl get settings.harvesterhci.io vm-migration-network -o yaml
    ```

    Example:

    ```yaml
    apiVersion: harvesterhci.io/v1beta1
    kind: Setting
    metadata:
      annotations:
        vm-migration-network.settings.harvesterhci.io/hash: ec8322fb6b741f94739cbb904fc73c3fda864d6d
        vm-migration-network.settings.harvesterhci.io/net-attach-def: harvester-system/vm-migration-network-6flk7
      creationTimestamp: "2022-10-13T06:36:39Z"
      generation: 51

>       name: storage-network

Reviewer: Should be vm migration network instead of `storage-network`?

Author: Updated it. Thanks.

---
sidebar_position: 12
sidebar_label: VM Migration Network
title: "VM Migration Network"
---

If the user wishes to isolate VM migration traffic from the Kubernetes cluster network (i.e. the management network) or other cluster-wide workloads. Users can allocate a dedicated vm migration network to get better network bandwidth and performance.

:::note

- Avoid configuring KubeVirt configuration directly, as this can result in unexpected or unwanted system behavior.

:::

> ## Prerequisites
> 
> There are some prerequisites before configuring the Harvester VM Migration Network setting.
> 
> - Well-configured Cluster Network and VLAN Config.
>     - Users have to ensure the Cluster Network is configured and VLAN Config will cover all nodes and ensure the network connectivity is working and expected in all nodes.
> - No VM Migration in progress before configuring the VM Migration Network setting.
> 
> :::caution
> 
> If the Harvester cluster was upgraded from v1.0.3, please check if Whereabouts CNI is installed properly before you move on to the next step. We will always recommend following this guide to check. [Issue 3168](https://github.com/harvester/harvester/issues/3168) describes that the Harvester cluster will not always install Whereabouts CNI properly.
> 
> - Verify the `ippools.whereabouts.cni.cncf.io` CRD exists with the following command.
>     - `kubectl get crd ippools.whereabouts.cni.cncf.io`
> 
> :::

Reviewer 2: ```suggestion
## Prerequisites

Before you begin configuring the VM migration network, ensure that the following requirements are met:

- The network switches are correctly configured, and a dedicated VLAN ID is assigned to the VM migration network.

- The [cluster network](../networking/clusternetwork.md) and [VLAN network](../networking/harvester-network.md) are configured correctly. Ensure that both networks cover all nodes and are accessible.

- No virtual machines are being migrated.

- The `ippools.whereabouts.cni.cncf.io` CRD exists. You can check this using the command `kubectl get crd ippools.whereabouts.cni.cncf.io`. In certain [upgrade scenarios](https://github.com/harvester/harvester/issues/3168), the Whereabouts CNI is not installed correctly.
```

> ## Configuration Example
> 
> - VLAN ID
> 	- Please check with your network switch setting, and provide a dedicated VLAN ID for VM Migration Network.
> - Well-configured Cluster Network and VLAN Config
> 	- Please refer Networking page for more details and configure [Cluster Network](../networking/clusternetwork.md) and [VLAN Config](../networking/harvester-network.md).
> - IP range for VM Migration Network
> 	- IP range should not conflict or overlap with Kubernetes cluster networks(`10.42.0.0/16`, `10.43.0.0/16`, `10.52.0.0/16` and `10.53.0.0/16` are reserved).
> 	- IP range should be in IPv4 CIDR format.
>   - Exclude IP addresses that KubeVirt pods and the VM migration network must not use.

Reviewer 2: ```suggestion
- The IP range of the VM migration network is in the IPv4 CIDR format and must neither conflict nor overlap with Kubernetes cluster networks. You must exclude IP addresses that KubeVirt pods and the VM migration network must not use. The following addresses are reserved: `10.42.0.0/16`, `10.43.0.0/16`, `10.52.0.0/16` and `10.53.0.0/16`.
```

Reviewer 2: The information must be part of the "Prerequisites" section. I will reorganize the "Storage Network" page after we merge this PR.

We will take the following configuration as an example to explain the details of the VM Migration Network

- VLAN ID for VM Migration Network: `100`
- Cluster Network: `vm-migration`
- IP range: `192.168.1.0/24`

> - Exclude Address: `192.168.1.1/32`

Reviewer 2: Please remove lines 43 to 48. The information is already mentioned in the CLI section, where it is relevant.

> ### Harvester VM Migration Network Setting
> 
> The [`vm-migration-network` setting](./settings.md#vm-migration-network) allows you to configure the network used to isolate in-cluster VM migration traffic when segregation is required.

Reviewer 2: ```suggestion
### `vm-migration-network` Setting

The [`vm-migration-network`](./settings.md#vm-migration-network) setting allows you to configure the network used to isolate in-cluster VM migration traffic when segregation is required.
```

You can [enable](#enable-the-vm-migration-network) and [disable](#disable-the-vm-migration-network) the VM migration network using either the UI or the CLI. When the setting is enabled, you must construct a Multus `NetworkAttachmentDefinition` CRD by configuring certain fields.

> #### Web UI
> 
> :::tip
> 
> Using the Harvester UI to configure the `vm-migration-network` setting is strongly recommended.
> 
> :::
> 
> ##### Enable the VM Migration Network
> 
> 1. Go to **Advanced > Settings > vm-migration-network**.
> 
> 1. Select **Enabled**.
> 
> 1. Configure the **VLAN ID**, **Cluster Network**, **IP Range**, and **Exclude** fields to construct a Multus `NetworkAttachmentDefinition` CRD.
> 
> 1. Click **Save**.
> 
> ![storage-network-enabled.png](/img/v1.4/storagenetwork/storage-network-enabled.png)
> 
> ##### Disable the VM Migration Network
> 
> 1. Go to **Advanced > Settings > vm-migration-network**.
> 
> 1. Select **Disabled**.
> 
> 1. Click **Save**.
> 
> Once the VM migration network is disabled, KubeVirt starts using the mgmt network for VM migration related operations.
> 
> ![storage-network-disabled.png](/img/v1.4/storagenetwork/storage-network-disabled.png)

Reviewer 2: ```suggestion
<Tabs>
<TabItem value="ui" label="UI" default>

:::tip

Using the Harvester UI to configure the `vm-migration-network` setting is strongly recommended.

:::

#### Enable the VM Migration Network

1. Go to **Advanced > Settings > vm-migration-network**.

1. Select **Enabled**.

1. Configure the **VLAN ID**, **Cluster Network**, **IP Range**, and **Exclude** fields to construct a Multus `NetworkAttachmentDefinition` CRD.

1. Click **Save**.

![storage-network-enabled.png](/img/v1.4/storagenetwork/storage-network-enabled.png)

#### Disable the VM Migration Network

1. Go to **Advanced > Settings > vm-migration-network**.

1. Select **Disabled**.

1. Click **Save**.

Once the VM migration network is disabled, KubeVirt starts using `mgmt` for VM migration-related operations.

![storage-network-disabled.png](/img/v1.4/storagenetwork/storage-network-disabled.png)

</TabItem>
```

Reviewer 2: Let's use tabs to reduce the number of headings and improve the reading experience.

> #### CLI
> 
> You can use the following command to configure the [`vm-migration-network` setting](./settings.md#vm-migration-network).
> 
> ```bash
> kubectl edit settings.harvesterhci.io vm-migration-network
> ```
> 
> The value format is JSON string or empty string as shown in below:
> 
> ```json
> {
>     "vlan": 100,
>     "clusterNetwork": "vm-migration",
>     "range": "192.168.1.0/24",
>     "exclude":[
>       "192.168.1.100/32"
>     ]
> }
> ```
> 
> The full configuration is like this example:
> 
> ```yaml
> apiVersion: harvesterhci.io/v1beta1
> kind: Setting
> metadata:
>   name: vm-migration-network
> value: '{"vlan":100,"clusterNetwork":"vm-migration","range":"192.168.1.0/24", "exclude":["192.168.1.100/32"]}'
> ```

Reviewer 2: ````suggestion
<TabItem value="cli" label="CLI">

You can use the following command to configure the [`vm-migration-network`](./settings.md#vm-migration-network) setting.

```bash
kubectl edit settings.harvesterhci.io vm-migration-network
```

The VM migration network is automatically enabled in the following situations:

- The value field contains a valid JSON string.

  ```yaml
  apiVersion: harvesterhci.io/v1beta1
  kind: Setting
  metadata:
    name: vm-migration-network
  value: '{"vlan":100,"clusterNetwork":"vm-migration","range":"192.168.1.0/24", "exclude":["192.168.1.100/32"]}'
  ```

- The value field is empty.

  ```yaml
  apiVersion: harvesterhci.io/v1beta1
  kind: Setting
  metadata:
    name: vm-migration-network
  value: ''
  ```
````

> When the VM migration network is disabled, the full configuration is as follows:
> 
> ```yaml
> apiVersion: harvesterhci.io/v1beta1
> kind: Setting
> metadata:
>   name: vm-migration-network
> ```
> 
> :::caution
> 
> Harvester considers extra insignificant characters in a JSON string as a different configuration.
> 
> Specifying a valid value in the `value` field enables the storage network. Deleting the `value` field disables the storage network.
> 
> :::

Reviewer 2: ````suggestion
The storage network is disabled when you remove the value field.

  ```yaml
  apiVersion: harvesterhci.io/v1beta1
  kind: Setting
  metadata:
    name: vm-migration-network
  ```

:::caution

Harvester considers extra insignificant characters in a JSON string as a different configuration.

:::

</TabItem>
</Tabs>
````

> ### After Applying Harvester VM Migration Network Setting
> 
> Harvester will create a new NetworkAttachmentDefinition and update the KubeVirt configuration.
> 
> Once the KubeVirt configuration is updated, KubeVirt will restart all `virt-handler` pods to apply the new network configuration.

Reviewer 2: ```suggestion
The following occur once the `vm-migration-network` setting is applied:

- Harvester creates a new `NetworkAttachmentDefinition` and updates the KubeVirt configuration.
- KubeVirt restarts all `virt-handler` pods to apply the new network configuration.
```

Reviewer 2: The information is part of the previous section. Also, the heading is not necessary because I reorganized this section.

> ### Verify Configuration is Completed
> 
> #### Step 1
> 
> Check if Harvester VM Migration Network setting's status is `True` and the type is `configured`.
> 
> ```bash
> kubectl get settings.harvesterhci.io vm-migration-network -o yaml
> ```
> 
> Completed Setting Example:
> 
> ```yaml
> apiVersion: harvesterhci.io/v1beta1
> kind: Setting
> metadata:
>   annotations:
>     vm-migration-network.settings.harvesterhci.io/hash: ec8322fb6b741f94739cbb904fc73c3fda864d6d
>     vm-migration-network.settings.harvesterhci.io/net-attach-def: harvester-system/vm-migration-network-6flk7
>   creationTimestamp: "2022-10-13T06:36:39Z"
>   generation: 51
>   name: storage-network
>   resourceVersion: "154638"
>   uid: 2233ad63-ee52-45f6-a79c-147e48fc88db
> status:
>   conditions:
>   - lastUpdateTime: "2022-10-13T13:05:17Z"
>     reason: Completed
>     status: "True"
>     type: configured
> ```

Reviewer 2: ````suggestion
### Post-Configuration Steps

1. Verify that the setting's status is `True` and the type is `configured` using the following command:

    ```bash
    kubectl get settings.harvesterhci.io vm-migration-network -o yaml
    ```

    Example:

    ```yaml
    apiVersion: harvesterhci.io/v1beta1
    kind: Setting
    metadata:
      annotations:
        vm-migration-network.settings.harvesterhci.io/hash: ec8322fb6b741f94739cbb904fc73c3fda864d6d
        vm-migration-network.settings.harvesterhci.io/net-attach-def: harvester-system/vm-migration-network-6flk7
      creationTimestamp: "2022-10-13T06:36:39Z"
      generation: 51
      name: storage-network
      resourceVersion: "154638"
      uid: 2233ad63-ee52-45f6-a79c-147e48fc88db
    status:
      conditions:
      - lastUpdateTime: "2022-10-13T13:05:17Z"
        reason: Completed
        status: "True"
        type: configured
    ```
````

> #### Step 2
> 
> Verify the readiness of all KubeVirt `virt-handler` pods, and confirm that their networks are correctly configured.
> 
> Execute the following command to inspect a pod's details:
> 
> ```bash
> kubectl -n harvester-system describe pod <pod-name>
> ```

Reviewer 2: ````suggestion
1. Verify that all KubeVirt `virt-handler` pods are ready and that their networks are correctly configured.

    You can inspect pod details using the following command:

    ```bash
    kubectl -n harvester-system describe pod <pod-name>
    ```
````

> #### Step 3
> 
> Check the `k8s.v1.cni.cncf.io/network-status` annotations and ensure that an interface named `migration0` exists, with an IP address within the designated IP range.
> 
> Users could use the following command to show all `virt-handler` pods to verify.
> 
> ```bash
> kubectl get pods -n harvester-system -l kubevirt.io=virt-handler -o yaml
> ```
> 
> Correct Network Example:
> 
> ```yaml
> apiVersion: v1
> kind: Pod
> metadata:
>   annotations:
>     cni.projectcalico.org/containerID: 004522bc8468ea707038b43813cce2fba144f0e97551d2d358808d57caf7b543
>     cni.projectcalico.org/podIP: 10.52.2.122/32
>     cni.projectcalico.org/podIPs: 10.52.2.122/32
>     k8s.v1.cni.cncf.io/network-status: |-
>       [{
>           "name": "k8s-pod-network",
>           "ips": [
>               "10.52.2.122"
>           ],
>           "default": true,
>           "dns": {}
>       },{
>           "name": "harvester-system/vm-migration-network-6flk7",
>           "interface": "migration0",
>           "ips": [
>               "10.1.2.1"
>           ],
>           "mac": "c6:30:6f:02:52:3e",
>           "dns": {}
>       }]
>     k8s.v1.cni.cncf.io/networks: vm-migration-network-6flk7@migration0
> 
> Omitted...
> ```

Reviewer 2: ````suggestion
1. Check the `k8s.v1.cni.cncf.io/network-status` annotations and verify that an interface named `migration0` exists. The IP address of this interface must be within the designated IP range.

    You can retrieve a list of `virt-handler` pods using the following command:

    ```bash
    kubectl get pods -n harvester-system -l kubevirt.io=virt-handler -o yaml
    ```

    Example:

    ```yaml
    apiVersion: v1
    kind: Pod
    metadata:
      annotations:
        cni.projectcalico.org/containerID: 004522bc8468ea707038b43813cce2fba144f0e97551d2d358808d57caf7b543
        cni.projectcalico.org/podIP: 10.52.2.122/32
        cni.projectcalico.org/podIPs: 10.52.2.122/32
        k8s.v1.cni.cncf.io/network-status: |-
          [{
              "name": "k8s-pod-network",
              "ips": [
                  "10.52.2.122"
              ],
              "default": true,
              "dns": {}
          },{
              "name": "harvester-system/vm-migration-network-6flk7",
              "interface": "migration0",
              "ips": [
                  "10.1.2.1"
              ],
              "mac": "c6:30:6f:02:52:3e",
              "dns": {}
          }]
        k8s.v1.cni.cncf.io/networks: vm-migration-network-6flk7@migration0
    Omitted...
    ```
````

## Best Practices

> - When configuring an [IP range](#configuration-example) for the VM migration network, ensure that the allocated IP addresses can service the future needs of the cluster. This is important because KubeVirt pods (`virt-handler`) stop running when new nodes are added to the cluster after the VM migration network is configured, and when the required number of IPs exceeds the allocated IPs. Resolving the issue involves reconfiguring the storage network with the correct IP range.

Reviewer 2: ```suggestion
- When configuring an [IP range](#prerequisites) for the VM migration network, ensure that the allocated IP addresses can service the future needs of the cluster. This is important because KubeVirt pods (`virt-handler`) stop running when new nodes are added to the cluster after the VM migration network is configured, and when the required number of IPs exceeds the allocated IPs. Resolving the issue involves reconfiguring the storage network with the correct IP range.
```

:::note

- Avoid configuring KubeVirt configuration directly, as this can result in unexpected or unwanted system behavior.

:::

## Prerequisites

There are some prerequisites before configuring the Harvester VM Migration Network setting.

- Well-configured Cluster Network and VLAN Config.
    - Users have to ensure the Cluster Network is configured and VLAN Config will cover all nodes and ensure the network connectivity is working and expected in all nodes.
- No VM Migration in progress before configuring the VM Migration Network setting.

:::caution

If the Harvester cluster was upgraded from v1.0.3, please check if Whereabouts CNI is installed properly before you move on to the next step. We will always recommend following this guide to check. [Issue 3168](https://github.com/harvester/harvester/issues/3168) describes that the Harvester cluster will not always install Whereabouts CNI properly.

- Verify the `ippools.whereabouts.cni.cncf.io` CRD exists with the following command.
    - `kubectl get crd ippools.whereabouts.cni.cncf.io`

> - If the Harvester cluster doesn't have `ippools.whereabouts.cni.cncf.io`, please add [these two CRDs](https://github.com/harvester/harvester/tree/v1.1.0/deploy/charts/harvester/dependency_charts/whereabouts/crds) before configuring `storage-network` setting.

Reviewer: Why suggest using the CRDs in 1.1.0 instead of the corresponding Harvester release? Might just copy from the storage network?

Author: Remove this since we already have ippools in Harvester.

```
kubectl apply -f https://raw.githubusercontent.com/harvester/harvester/v1.1.0/deploy/charts/harvester/dependency_charts/whereabouts/crds/whereabouts.cni.cncf.io_ippools.yaml
kubectl apply -f https://raw.githubusercontent.com/harvester/harvester/v1.1.0/deploy/charts/harvester/dependency_charts/whereabouts/crds/whereabouts.cni.cncf.io_overlappingrangeipreservations.yaml
```

:::

## Configuration Example

- VLAN ID
	- Please check with your network switch setting, and provide a dedicated VLAN ID for VM Migration Network.
- Well-configured Cluster Network and VLAN Config

> 	- Please refer Networking page for more details and configure `Cluster Network` and `VLAN Config` but not `Networks`.

Reviewer: Suggest adding the corresponding doc for reference.

- IP range for VM Migration Network
	- IP range should not conflict or overlap with Kubernetes cluster networks(`10.42.0.0/16`, `10.43.0.0/16`, `10.52.0.0/16` and `10.53.0.0/16` are reserved).
	- IP range should be in IPv4 CIDR format.

>   - Exclude IP addresses that KubeVirt pods and the VM migration network must not use.

Reviewer: Wrong indent? 
What IP addresses should KubeVirt not use?

Author: This follows content in storage networks.

https://docs.harvesterhci.io/v1.6/advanced/storagenetwork/#configuration-example
